Gallery HTML build broken after upgrade vom 0.11.1 to 0.12.0

My nightly doc-builds of multiple projects suddenly failed without any change to the software projects themself and I tracked it down to a single change between the last successful build and the current failing builds: the upgrade of `sphinx-gallery` from v0.11.1 to 0.12.0.

I am currently running these versions under Python 3.10.9:
```
$ pip freeze | grep -i sphinx
Sphinx==6.1.3
sphinx-autoapi==2.0.1
sphinx-gallery==0.12.0
sphinx-rtd-theme==0.5.1
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.0
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
sphinxcontrib-versioning==2.2.1
sphinxcontrib.applehelp==1.0.3
```

The initial error log in the CI was:

```
[AutoAPI] Reading files... [ 95%] /builds/km3py/km3io/src/km3io/_definitions/applications.py
[AutoAPI] Reading files... [100%] /builds/km3py/km3io/src/km3io/utils/kprinttree.py
[AutoAPI] Mapping Data... [  4%] /builds/km3py/km3io/src/km3io/rootio.py
Extension error (sphinx_gallery.gen_gallery):
Handler <function write_api_entries at 0x7f4e63070940> for event 'autodoc-process-docstring' threw an exception (exception: 'show_api_usage')
make: *** [Makefile:20: html] Error 2

Cleaning up project directory and file based variables
00:01
ERROR: Job failed: exit code 1
```

and after reproducing it locally, I ran the command with more verbosity and got this:

```python
[app] emitting event: 'autodoc-process-docstring'('method', '__init__', None, None, ['Initialize self.  See help(type(self)) for accurate signature.'

Traceback (most recent call last):
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/sphinx/events.py", line 96, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/sphinx_gallery/gen_gallery.py", line 743, in write_api_entries
    if app.config.sphinx_gallery_conf['show_api_usage'] is False:
KeyError: 'show_api_usage'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/sphinx/cmd/build.py", line 279, in build_main
    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/sphinx/application.py", line 262, in __init__
    self._init_builder()
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/sphinx/application.py", line 335, in _init_builder
    self.events.emit('builder-inited')
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/sphinx/events.py", line 96, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/autoapi/extension.py", line 156, in run_autoapi
    sphinx_mapper_obj.map(options=app.config.autoapi_options)
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/autoapi/mappers/python/mapper.py", line 343, in map
    super().map(options)
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/autoapi/mappers/base.py", line 297, in map
    for obj in self.create_class(data, options=options):
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/autoapi/mappers/python/mapper.py", line 382, in create_class
    for child_obj in self.create_class(
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/autoapi/mappers/python/mapper.py", line 382, in create_class
    for child_obj in self.create_class(
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/autoapi/mappers/python/mapper.py", line 395, in create_class
    self.app.emit(
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/sphinx/application.py", line 458, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/sphinx/events.py", line 107, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function write_api_entries at 0x11001fac0> for event 'autodoc-process-docstring' threw an exception (exception: 'show_api_usage')

Extension error (sphinx_gallery.gen_gallery):
Handler <function write_api_entries at 0x11001fac0> for event 'autodoc-process-docstring' threw an exception (exception: 'show_api_usage')
```

So apparently it chokes on a missing entry in the `sphinx_gallery_conf` named `'show_api_usage'`.

```
    if app.config.sphinx_gallery_conf['show_api_usage'] is False:
KeyError: 'show_api_usage'
```

I added that in my `conf.py`:

```python
sphinx_gallery_conf = {
    "backreferences_dir": "modules/generated",
    "default_thumb_file": "_static/default_gallery_thumbnail.png",
    "examples_dirs": "../examples",  # path to your example scripts
    "gallery_dirs": "auto_examples",  # path to where to save gallery generated output
    "show_memory": True,
    "show_api_usage": True,
}
```

but then got this:

```
[AutoAPI] Rendering Data... [ 95%] km3io._definitions.trigger
[AutoAPI] Rendering Data... [100%] km3io._definitions.daqdatatypes

generating gallery...

Configuration error:
Unknown key(s) in sphinx_gallery_conf:
'api_entries'
make: *** [html] Error 2
```

changing it to `False` threw another `Extension error`:

```
[AutoAPI] Rendering Data... [ 95%] km3io._definitions.trigger
[AutoAPI] Rendering Data... [100%] km3io._definitions.daqdatatypes

generating gallery...

Extension error (sphinx_gallery.gen_gallery):
Handler <function generate_gallery_rst at 0x104143400> for event 'builder-inited' threw an exception (exception: process no longer exists (pid=77972))
make: *** [html] Error 2
```

So I cranked up the verbosity again with `sphinx-build -M html "." "_build" -vvv` and found another problem related to `psutil` (installed version is `psutil==5.9.4`, but it works with `sphinx-gallery==0.11.1`):

```
Traceback (most recent call last):
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/psutil/_common.py", line 443, in wrapper
    ret = self._cache[fun]
AttributeError: 'Process' object has no attribute '_cache'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/psutil/_common.py", line 443, in wrapper
    ret = self._cache[fun]
AttributeError: 'Process' object has no attribute '_cache'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/psutil/_psosx.py", line 346, in wrapper
    return fun(self, *args, **kwargs)
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/psutil/_common.py", line 446, in wrapper
    return fun(self)
  File "/Users/tamasgal/Dev/km3io/venv/lib/python3.10/site-packages/psutil/_psosx.py", line 381, in _get_pidtaskinfo
    ret = cext.proc_pidtaskinfo_oneshot(self.pid)
ProcessLookupError: [Errno 3] No such process (originated from proc_pidinfo())
```

Any ideas?

--- a/sphinx_gallery/gen_gallery.py
+++ b/sphinx_gallery/gen_gallery.py
@@ -158,7 +158,10 @@ def _complete_gallery_conf(sphinx_gallery_conf, src_dir, plot_gallery,
                            builder_name='html', app=None, check_keys=True):
     gallery_conf = copy.deepcopy(DEFAULT_GALLERY_CONF)
     options = sorted(gallery_conf)
-    extra_keys = sorted(set(sphinx_gallery_conf) - set(options))
+    ignore_keys = ('_sg_api_entries',)  # sphinx-autoapi can cause us to add this
+    extra_keys = sorted(
+        (set(sphinx_gallery_conf) - set(options)) - set(ignore_keys)
+    )
     if extra_keys and check_keys:
         msg = 'Unknown key(s) in sphinx_gallery_conf:\n'
         for key in extra_keys:
@@ -737,11 +740,11 @@ def write_api_entries(app, what, name, obj, options, lines):
         app.config.sphinx_gallery_conf['show_api_usage'] = False
     if app.config.sphinx_gallery_conf['show_api_usage'] is False:
         return
-    if 'api_entries' not in app.config.sphinx_gallery_conf:
-        app.config.sphinx_gallery_conf['api_entries'] = dict()
-    if what not in app.config.sphinx_gallery_conf['api_entries']:
-        app.config.sphinx_gallery_conf['api_entries'][what] = set()
-    app.config.sphinx_gallery_conf['api_entries'][what].add(name)
+    if '_sg_api_entries' not in app.config.sphinx_gallery_conf:
+        app.config.sphinx_gallery_conf['_sg_api_entries'] = dict()
+    if what not in app.config.sphinx_gallery_conf['_sg_api_entries']:
+        app.config.sphinx_gallery_conf['_sg_api_entries'][what] = set()
+    app.config.sphinx_gallery_conf['_sg_api_entries'][what].add(name)
 
 
 def init_api_usage(gallery_dir):
@@ -812,7 +815,7 @@ def _make_graph(fname, entries, gallery_conf):
                 dg.edge(node_from, node_to)
         # add modules with all API entries
         dg.attr('node', color='lightblue2')
-        for module in gallery_conf['api_entries']['module']:
+        for module in gallery_conf['_sg_api_entries']['module']:
             struct = module.split('.')
             for i in range(len(struct) - 1):
                 if struct[i + 1] not in lut:
@@ -852,27 +855,27 @@ def write_api_entry_usage(app, docname, source):
     # to in a gallery directory, e.g. auto_examples), it runs last
     # which means that all the api entries will be in gallery_conf
     if 'sg_api_usage' not in docname or \
-            'api_entries' not in gallery_conf or \
+            '_sg_api_entries' not in gallery_conf or \
             gallery_conf['backreferences_dir'] is None:
         return
     backreferences_dir = os.path.join(gallery_conf['src_dir'],
                                       gallery_conf['backreferences_dir'])
 
     example_files = set.union(
-        *[gallery_conf['api_entries'][obj_type]
+        *[gallery_conf['_sg_api_entries'][obj_type]
           for obj_type in ('class', 'method', 'function')
-          if obj_type in gallery_conf['api_entries']])
+          if obj_type in gallery_conf['_sg_api_entries']])
 
     if len(example_files) == 0:
         return
 
     def get_entry_type(entry):
-        if entry in gallery_conf['api_entries'].get('class', []):
+        if entry in gallery_conf['_sg_api_entries'].get('class', []):
             return 'class'
-        elif entry in gallery_conf['api_entries'].get('method', []):
+        elif entry in gallery_conf['_sg_api_entries'].get('method', []):
             return 'meth'
         else:
-            assert entry in gallery_conf['api_entries']['function']
+            assert entry in gallery_conf['_sg_api_entries']['function']
             return 'func'
 
     # find used and unused API entries

--- a/sphinx_gallery/gen_gallery.py
+++ b/sphinx_gallery/gen_gallery.py
@@ -108,7 +108,7 @@ DEFAULT_GALLERY_CONF = {
     'nested_sections': True,
     'prefer_full_module': [],
     'api_usage_ignore': '.*__.*__',
-    'show_api_usage': False,
+    'show_api_usage': False,  # if this changes, change write_api_entries, too
     'copyfile_regex': '',
 }
 
@@ -732,10 +732,13 @@ def write_computation_times(gallery_conf, target_dir, costs):
 
 
 def write_api_entries(app, what, name, obj, options, lines):
-    if app.config.sphinx_gallery_conf['show_api_usage'] is False:
-        return
+    # sphinx-autoapi can do this before we manage to complete our config
+    if 'show_api_usage' not in app.config.sphinx_gallery_conf:
+        app.config.sphinx_gallery_conf['show_api_usage'] = False
     if 'api_entries' not in app.config.sphinx_gallery_conf:
         app.config.sphinx_gallery_conf['api_entries'] = dict()
+    if app.config.sphinx_gallery_conf['show_api_usage'] is False:
+        return
     if what not in app.config.sphinx_gallery_conf['api_entries']:
         app.config.sphinx_gallery_conf['api_entries'][what] = set()
     app.config.sphinx_gallery_conf['api_entries'][what].add(name)

--- a/sphinx_gallery/gen_gallery.py
+++ b/sphinx_gallery/gen_gallery.py
@@ -158,7 +158,11 @@ def _complete_gallery_conf(sphinx_gallery_conf, src_dir, plot_gallery,
                            builder_name='html', app=None, check_keys=True):
     gallery_conf = copy.deepcopy(DEFAULT_GALLERY_CONF)
     options = sorted(gallery_conf)
-    ignore_keys = ('_sg_api_entries',)  # sphinx-autoapi can cause us to add this
+    # sphinx-autoapi can run before our config filling can occur, which
+    # when used with show_api_usage=True can cause us to add this early.
+    # In theory we could bump our priority above theirs, but this could have
+    # unintented consequences.
+    ignore_keys = ('_sg_api_entries',)
     extra_keys = sorted(
         (set(sphinx_gallery_conf) - set(options)) - set(ignore_keys)
     )

--- a/sphinx_gallery/gen_gallery.py
+++ b/sphinx_gallery/gen_gallery.py
@@ -735,10 +735,10 @@ def write_api_entries(app, what, name, obj, options, lines):
     # sphinx-autoapi can do this before we manage to complete our config
     if 'show_api_usage' not in app.config.sphinx_gallery_conf:
         app.config.sphinx_gallery_conf['show_api_usage'] = False
-    if 'api_entries' not in app.config.sphinx_gallery_conf:
-        app.config.sphinx_gallery_conf['api_entries'] = dict()
     if app.config.sphinx_gallery_conf['show_api_usage'] is False:
         return
+    if 'api_entries' not in app.config.sphinx_gallery_conf:
+        app.config.sphinx_gallery_conf['api_entries'] = dict()
     if what not in app.config.sphinx_gallery_conf['api_entries']:
         app.config.sphinx_gallery_conf['api_entries'][what] = set()
     app.config.sphinx_gallery_conf['api_entries'][what].add(name)

